app/jobs/allocation_job.rb

```diff
@@ -2,38 +2,89 @@ class AllocationJob < ApplicationJob
+    set_negative_allocation_and_cap_values if @allocation_delta.negative?
 
-    current_cap_value = current_allocation&.raw_cap || 0
-    new_cap_value = [current_cap_value + allocation_batch_job.allocation_delta, devices_ordered].max
+    ActiveRecord::Base.transaction do
+      update_current_allocation!
+      call_school_order_state_and_cap_update_service
+      record_batch_job_processed_and_notify
+    end
+  end
 
-    disable_user_notifications = !allocation_batch_job.send_notification
+private
 
-    allocation = SchoolDeviceAllocation.find_or_initialize_by(school: school, device_type: 'std_device')
+  def set_allocation_batch_job
+    @allocation_batch_job = allocation_batch_job
+  end
 
-    ActiveRecord::Base.transaction do
-      allocation.update!(allocation: new_allocation_value, cap: new_cap_value)
-
-      service = SchoolOrderStateAndCapUpdateService.new(
-        school: school,
-        order_state: order_state,
-        std_device_cap: new_cap_value,
-      )
-
-      service.disable_user_notifications! if disable_user_notifications
-      service.call
-
-      if disable_user_notifications
-        allocation_batch_job.update!(processed: true)
-      else
-        allocation_batch_job.update!(processed: true, sent_notification: true)
-      end
+  def set_instance_variables(allocation_batch_job)
+    @allocation_batch_job = allocation_batch_job
+    @school = @allocation_batch_job.school
+    @order_state = @allocation_batch_job.order_state
+    @allocation_delta = @allocation_batch_job.allocation_delta.to_i
+    @disable_user_notifications = !@allocation_batch_job.send_notification
+    @current_allocation = SchoolDeviceAllocation.find_or_initialize_by(school: @school, device_type: 'std_device')
+  end
+
+  def set_default_new_raw_allocation_value
+    @current_raw_allocation_value = @current_allocation&.raw_allocation || 0
+    @new_raw_allocation_value = @current_raw_allocation_value + @allocation_delta
+  end
+
+  def set_default_new_raw_cap_value
+    @current_raw_cap_value = @current_allocation&.raw_cap || 0
+    @new_raw_cap_value = @current_raw_cap_value + @allocation_delta
+  end
+
+  def set_negative_allocation_and_cap_values
+    set_negative_allocation_delta
+    set_negative_new_raw_allocation_value
+    set_negative_new_raw_cap_value
+  end
+
+  def set_negative_allocation_delta
+    @negative_allocation_delta = [-@current_allocation.devices_available_to_order, @allocation_delta].max
+  end
+
+  def set_negative_new_raw_allocation_value
+    @new_raw_allocation_value = @current_raw_allocation_value + @negative_allocation_delta
+  end
+
+  def set_negative_new_raw_cap_value
+    @new_raw_cap_value = [@current_raw_cap_value + @negative_allocation_delta, @new_raw_allocation_value].min
+  end
+
+  def update_current_allocation!
+    if @allocation_delta.negative?
+      return unless @current_allocation.devices_available_to_order.positive? # Don't update the allocation, but allow order_state to be updated
```

Comment:
May be worth using `@current_allocation.devices_available_to_order?`

Reply:
Thinking about this change some more: is it worth decreasing the allocation amount even if no devices are available to order? I guess it depends on the _intent_ of this batch job – should it "safeguard" allocation (and thus cap) amounts, or should it make the intended change regardless (and then assume the rest of the system can handle the updated allocation amount regardless of the devices available to order)?